[clang] Keep using declarations at head of list in addOrReplaceDecl

nicovank opened the pull request:
Fix #104883.
Fix #104800.

Thanks @Endilll for the reduced reproducer. Bisection points to 0cb7e7ca0c864e052bf49978f3bcd667c9e16930 as the faulty commit, but nothing there seems to indicate what the problem is. After some print debugging, `Decl` on [this line](https://github.com/llvm/llvm-project/blob/f22b1da8791edd557ce34c87190e329df2e1c892/clang/lib/Sema/SemaOverload.cpp#L7429) is a `UsingDecl`, not a `CXXMethodDecl`, and the cast fails.

Add a conditional case for `UsingDecl`.

I am not familiar with Clang/Sema so please make sure this fix makes sense.

mizvekov commented on clang/lib/Sema/SemaOverload.cpp:
```diff
@@ -7425,6 +7425,13 @@ void Sema::AddMethodCandidate(DeclAccessPair FoundDecl, QualType ObjectType,
                                ObjectClassification, Args, CandidateSet,
                                SuppressUserConversions, false, PO);
+  } else if (UsingDecl *UD = dyn_cast<UsingDecl>(Decl)) {
+    for (auto *Shadow : UD->shadows()) {
+      AddMethodCandidate(cast<CXXMethodDecl>(Shadow->getTargetDecl()),
+                         FoundDecl, ActingContext, ObjectType,
+                         ObjectClassification, Args, CandidateSet,
+                         SuppressUserConversions, false, std::nullopt, PO);
```
I think this needs better investigation why that commit introduced this change.

One concerns is that the target decl could be a FunctionTemplate as well.

nicovank replied:
Thanks, correct, template operators still cause a crash. I'll spend a bit of time digging into the original problematic commit.

nicovank committed:
[clang] Keep using declarations at head of list in addOrReplaceDecl

Fix #104883.
Fix #104800.

0cb7e7ca0c864e052bf49978f3bcd667c9e16930 is the culprit by bisection. The reason
for the break is that prior to this commit, using declarations were always kept
at the front of the list. This behavior was changed in that commit.

Here, we re-establish the behavior of keeping using declarations at the head in
`addOrReplaceDecl`.

nicovank commented:
The reason for the break in 0cb7e7ca0c864e052bf49978f3bcd667c9e16930 is that prior to this commit, using declarations were always kept at the front of the `NamedDecl` list. This behavior was changed in that commit.

See a comment here:
https://github.com/llvm/llvm-project/blob/3b8b5d1f22f2c5d8ae55709a92d02011dea056ca/clang/include/clang/AST/DeclContextInternals.h#L213-L216

Here, we re-establish the behavior of keeping using declarations at the head in `addOrReplaceDecl`.
https://github.com/llvm/llvm-project/blob/3b8b5d1f22f2c5d8ae55709a92d02011dea056ca/clang/include/clang/AST/DeclContextInternals.h#L218-L224

Some thoughts:
 1. Seems to me like this order shouldn't be relied on somewhere else in the code. But I'm not familiar.
 2. I conjecture that this custom-built linked list is overkill and doesn't actually have any significant runtime or memory win over `std::forward_list` for example. It makes the code hard to work with. But it's also pretty deeply ingrained into the code at this point.

Let me know what you think of this change (seems like the right way to go / worth it?) or any other solutions. It's a little heavy at the moment with the triple `Decl::IDNS_Using` check, I'll think about simplifying it.
